### DISCUSSION TITLE
Tie specific BIOSAttrs over to SecurityAttrs

### DISCUSSION THREAD

**Discussion by superm1** [fwupd/fwupd#4880] [Ideas]
With `FwupdBiosAttr` merged in now, there are opportunities to use it to improve the experience someone gets with `FwupdSecurityAttr`.

If previously you would show a message like "Enter BIOS setup to turn on UEFI secure boot", you can now turn it on directly from fwupd.  Or if the IOMMU is turned off in BIOS setup, you can turn it on directly in fwupd.

So I was thinking it might make sense to add add a new member to `FwupdBiosAttr` for the appstream ID of a `FwupdSecurityAttr` that it can help improve.  This would need to be a manually built hardcoded heuristic table, but the number of things that can be changed to improve a given system is small enough that I don't think it will matter too much.

**Comment by superm1** [comment]
Of course the heuristic table can go the other way too to have a member of `FwupdSecurityAttr` for the `FwupdBiosAttr` that it is tied to.

**Reply by hughsie** [reply]
It think FwupdBiosAttr on a FwupdSecurityAttr makes sense. I am slightly hesitant to make it super easy with no scary warning, as if we (for instance) turned on IOMMU and then the integrated graphics then didn't initialize the user would have to navigate the system BIOS at boot time to _fix things_.

**Reply by superm1** [reply]
Sure. I think that's for the client to show such a scary warning though.

**Reply by hughsie** [reply]
> I think that's for the client to show such a scary warning though.

Ack, agree.

**Comment by hughsie** [comment]
I was looking at the code in more detail today, and I'm wondering if we've got the `preferred_value` on the right object. Semantically it's the desire of the security attribute, not the bios attribute, i.e. you could imagine that a specific attribute would have to be "_foo or bar or baz_" for HSI-1 and "_only foo_" for HSI-2 -- and we have no way to express that right now. Would you be against me moving the `bios_preferred_value` to the `FwupdSecurityAttr` instead? Maybe it also makes sense to have something like:

    fwupd_security_attr_set_bios_attr(const gchar *id, const gchar *preferred_value)

or even....

    fu_security_attr_add_bios_fix(FwupdSecturityAttr *attr, const gchar *id, const gchar *preferred_value)

so you could do something like:

    fu_security_attr_add_bios_fix(attr, "AmdVt", "enable")
    fu_security_attr_add_bios_fix(attr, "IOMMU", "enable")
    fu_security_attr_add_bios_fix(attr, "VtForDirectIo", "enable")
    fu_security_attr_add_bios_fix(attr, "SomethingElse", "on")

...and then the daemon can match the key against any of the bios attrs, and also check if the value is one of the allowed enumerations. Sane?

**Reply by hughsie** [reply]
e.g. https://github.com/fwupd/fwupd/pull/4888
